### DISCUSSION TITLE
Find included files

### DISCUSSION THREAD

**Discussion by addw** [smarty-php/smarty#844] [Ideas]
I want to be able to get a list of all templates {include}d within another, something that I did with Smarty 2 by looking at $template->properties['file_dependency']. This should ideally include all {include} within an {include}.
I want to do this after running  $smarty->createTemplate() and generate a dependency file that lists the included files.
I can see some of this in the _templates_c/xxx.php file but need to do it in the script that calls createTemplate().

The reason for this is that I cache the HTML output and have Apache serve it up without involving PHP at all - so spotting that cached HTML is out of date has to be done by running a separate script which looks at the dependency file.

I suspect that this will be something new.

**Comment by addw** [comment]
I solved that problem using the code below (which writes the include paths to a file). 
$tpl I get from $smarty->createTemplate()

This also does not find includes within includes.

I am not very happy with the replacement of "file:" but it seems to work. Suggestions for doing better will be gratefully received.
```
// This should yield the main template
// file_dependency is an array of arrays where [0] is an absolute path
// [2] is a type, should be "file"
foreach($tpl->compiled->file_dependency as $dep) {
    // Make path relative
    $fpath = str_replace($_SERVER['DOCUMENT_ROOT'], ".", $dep[0]);
    fwrite($depfh, "r $fpath\n") or $werr = TRUE;
}
             
// includes in an array where the keys are like "file:_header.tpl"
// _joined_template_dir is an absolute path to the directory "_templates/"
// that contains the above main template
foreach($tpl->compiled->includes as $dep => $v) {
    $dep = str_replace("file:", $tpl->source->smarty->_joined_template_dir, $dep);
    $dep = str_replace($_SERVER['DOCUMENT_ROOT'], ".", $dep);
    fwrite($depfh, "r {$dep}\n") or $werr = TRUE;
}
```

**Reply by wisskid** [reply]
Why not just use isCached()?

**Comment by addw** [comment]
Summary: I cannot use isCached() as smarty is not always invoked.

All web pages are generated with smarty.
All pages include standard components, eg page header - these will rarely change.
Some web pages are dynamic, some are static (ie rarely change - template just contains HTML).
For efficiency the static pages are compiled to HTML and this HTML put somewhere that
the web server (Apache) can find and send to the browser without invoking PHP/smarty.
As smarty is not invoked isCached() cannot be used.

If a static page (or a component) changes the cached HTML needs to be marked Out-of-date
so that Apache will run PHP/smarty.
This is done by running a script that looks at the files on which the cached HTML depends.
This script needs to know what the dependencies are to be able to work.

**Reply by wisskid** [reply]
Use PHP in your script then? It runs perfectly from command line or cronjobs too.

**Reply by addw** [reply]
I shall give that a go when I have some time in a couple of weeks.
Thanks for the suggestion.
